Modules/_hashopenssl.c

```diff
@@ -1879,94 +1880,102 @@ hashlib_free(void *m)
+{
+    _hashlibstate *state = get_hashlib_state(module);
 
-    PyTypeObject *EVPtype = (PyTypeObject *)PyType_FromSpec(&EVPtype_spec);
-    if (EVPtype == NULL) {
-        Py_DECREF(m);
-        return NULL;
+    state->EVPtype = (PyTypeObject *)PyType_FromSpec(&EVPtype_spec);
+    if (state->EVPtype == NULL) {
+        return -1;
     }
-    state->EVPtype = EVPtype;
-    Py_INCREF((PyObject *)state->EVPtype);
-    PyModule_AddObject(m, "HASH", (PyObject *)state->EVPtype);
-
-    PyTypeObject *HMACtype = (PyTypeObject *)PyType_FromSpec(&HMACtype_spec);
-    if (HMACtype == NULL) {
-        Py_DECREF(m);
-        return NULL;
+    if (PyModule_AddType(module, state->EVPtype) < 0) {
+        return -1;
     }
-    state->HMACtype = HMACtype;
-    Py_INCREF((PyObject *)state->HMACtype);
-    PyModule_AddObject(m, "HMAC", (PyObject *)state->HMACtype);
+    return 0;
+}
 
+static int
+hashlib_init_evpxoftype(PyObject *module)
+{
 #ifdef PY_OPENSSL_HAS_SHAKE
-    bases = PyTuple_Pack(1, (PyObject *)EVPtype);
+    _hashlibstate *state = get_hashlib_state(module);
+    PyObject *bases;
+
+    if (state->EVPtype == NULL) {
+        return -1;
+    }
+
+    bases = PyTuple_Pack(1, state->EVPtype);
     if (bases == NULL) {
-        Py_DECREF(m);
-        return NULL;
+        return -1;
     }
-    PyTypeObject *EVPXOFtype = (PyTypeObject *)PyType_FromSpecWithBases(
+
+    state->EVPXOFtype = (PyTypeObject *)PyType_FromSpecWithBases(
         &EVPXOFtype_spec, bases
     );
     Py_DECREF(bases);
-    if (EVPXOFtype == NULL) {
-        Py_DECREF(m);
-        return NULL;
+    if (state->EVPXOFtype == NULL) {
+        return -1;
+    }
+    if (PyModule_AddType(module, state->EVPXOFtype) < 0) {
+        return -1;
     }
-    state->EVPXOFtype = EVPXOFtype;
-
-    Py_INCREF((PyObject *)state->EVPXOFtype);
-    PyModule_AddObject(m, "HASHXOF", (PyObject *)state->EVPXOFtype);
 #endif
+    return 0;
+}
 
-    openssl_md_meth_names = generate_hash_name_list();
-    if (openssl_md_meth_names == NULL) {
-        Py_DECREF(m);
-        return NULL;
+static int
+hashlib_init_hmactype(PyObject *module)
+{
+    _hashlibstate *state = get_hashlib_state(module);
+
+    state->HMACtype = (PyTypeObject *)PyType_FromSpec(&HMACtype_spec);
+    if (state->HMACtype == NULL) {
+        return -1;
     }
-    if (PyModule_AddObject(m, "openssl_md_meth_names", openssl_md_meth_names)) {
-        Py_DECREF(m);
-        return NULL;
+    if (PyModule_AddType(module, state->HMACtype) < 0) {
+        return -1;
     }
+    return 0;
+}
+
+static PyModuleDef_Slot hashlib_slots[] = {
+    /* OpenSSL 1.0.2 and LibreSSL */
+    {Py_mod_exec, hashlib_openssl_legacy_init},
+    {Py_mod_exec, hashlib_init_evptype},
+    {Py_mod_exec, hashlib_init_evpxoftype},
+    {Py_mod_exec, hashlib_init_hmactype},
+    {Py_mod_exec, hashlib_md_meth_names},
+    {0, NULL}
+};
 
-    PyState_AddModule(m, &_hashlibmodule);
-    return m;
+static struct PyModuleDef _hashlibmodule = {
```

Comment:
designated initializers (especially nice for type declarations) from pep7: https://www.python.org/dev/peps/pep-0007/

PS: I am not sure it  is optional or not~